### ISSUE TITLE
[question] Respecting compiler.libcxx abi setting

### ISSUE THREAD

**Issue by brevzin** [conan-io/conan#11958]
I have the following conan profile:

```
$ conan --version
Conan version 1.51.3

$ conan profile show default
Configuration for profile default:

[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=11
compiler.cppstd=20
compiler.libcxx=libstdc++
build_type=Release
[options]
[conf]
[build_requires]
[env]
```

Note that `compiler.libcxx` is the old ABI (`libstdc++`) and not the new ABI (`libstdc++11`). This is intentional. Also I'm specifying `cppstd=20`.

I then made a custom conanfile for `{fmt}`, which tries to package up fmt 9.0. That looks like, in its entirety:

```python
from conans import ConanFile, CMake, tools
from conan.tools.cmake import CMakeToolchain, cmake_layout

class ConanFmt(ConanFile):
    name = "fmt"
    version = "9.0.0"

    settings = "os", "compiler", "cppstd", "build_type", "arch"

    def source(self):
        git = tools.Git(folder="fmt")
        git.clone("https://github.com/fmtlib/fmt.git", "9.0.0", shallow=True)

    def layout(self):
        cmake_layout(self)

    def generate(self):
        tc = CMakeToolchain(self)
        tc.generate()

    def build(self):
        cmake = CMake(self)
        cmake.definitions["FMT_DOC"] = False
        cmake.definitions["FMT_TEST"] = False
        cmake.definitions["FMT_INSTALL"] = True
        cmake.definitions["CMAKE_EXPORT_COMPILE_COMMANDS"] = True
        cmake.configure(source_folder="fmt")
        cmake.build()

    def package(self):
        cmake = CMake(self)
        cmake.install()

    def package_info(self):
        self.cpp_info.libs = ["fmt"]
```

When I try to go ahead and package this up, like say:

```
$ conan create . demo/testing -s build_type=Release
```

This does actually go ahead and clone the `{fmt}` repo, and correctly build and package up the contents of the library.

Except that:

* there is no `_GLIBCXX_USE_CXX11_ABI` definition set (despite what [this page](https://docs.conan.io/en/latest/howtos/manage_gcc_abi.html#manage-gcc-abi)) says. 
* `{fmt}` was compiled with C++11, not C++20
* ... but it _was_ built with `Release` (`-O3 -DNDEBUG`)

Conan does appear to provide variables like `CONAN_CMAKE_CXX_STANDARD=20`, `CONAN_STD_CXX_FLAG=-std=c++2a`, and `CONAN_LIBCXX=libstdc++`. But `{fmt}`'s build doesn't actually use those, so that isn't very useful.

How do I actually get conan to propagate through these variables properly? Is that something that must be explicitly specified in the `conanfile.py` (even if it's in the profile)?

I also posted this question on [StackOverflow](https://stackoverflow.com/q/73476787/2069064).

**Comment by memsharded** [comment]
Hi @brevzin 

This would be a working "modern" recipe based on yours:
```python
from conan import ConanFile
from conan.tools.scm import Git
from conan.tools.cmake import CMakeToolchain, cmake_layout, CMake

class ConanFmt(ConanFile):
    name = "fmt"
    version = "9.0.0"

    settings = "os", "compiler", "build_type", "arch"

    def source(self):
        git = Git(self)
        git.clone("https://github.com/fmtlib/fmt.git", target=".")
        git.checkout(self.version)

    def layout(self):
        cmake_layout(self, src_folder="fmt")

    def generate(self):
        tc = CMakeToolchain(self)
        tc.variables["FMT_DOC"] = False
        tc.variables["FMT_TEST"] = False
        tc.variables["FMT_INSTALL"] = True
        tc.variables["CMAKE_EXPORT_COMPILE_COMMANDS"] = True
        tc.generate()

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def package(self):
        cmake = CMake(self)
        cmake.install()

    def package_info(self):
        self.cpp_info.libs = ["fmt"]
```

The changes are:
- Dropped global ``cppstd``, now it is a ``compiler.cppstd`` subsetting.
- Moved ``cmake.definitions`` to ``cmaketoolchain.variables``
- Used modern ``from conan.tools.scm import Git`` git helper, experimental, but 2.0-ready

This seems to work with ``conan create .``
A good starting point for recipes is use the ``conan new hello/0.1 --template=cmake_lib``, that will also include a ``test_package``.

**Comment by brevzin** [comment]
Great! That does seem to work, and propagate through the relevant pieces. I then also changed `package_info` to do... this:

```python
def package_info(self):
    self.cpp_info.libs = ["fmt" if self.settings.get_safe("build_type") == "Release" else "fmtd"]
```

since `{fmt}` emits a `libfmt.a` in release mode but `libfmtd.a` in debug mode, and that... also seems to work. Right way to do that?

---

This is also kind of a separate question, but what is then the right way to pass all this same conan profile information (libstdc++ version, c++ standard version, build type) into cmake? I see the generated `conanbuildinfo.cmake` that I get from running `conan install` does conditionally try to do the right thing:
```cmake
if(CONAN_LIBCXX STREQUAL "libstdc++11")
    add_definitions(-D_GLIBCXX_USE_CXX11_ABI=1)
elseif(CONAN_LIBCXX STREQUAL "libstdc++")
    add_definitions(-D_GLIBCXX_USE_CXX11_ABI=0)
endif()
```
except that `CONAN_LIBCXX` happens to not be set, so this doesn't happen.

**Comment by memsharded** [comment]
> since {fmt} emits a libfmt.a in release mode but libfmtd.a in debug mode, and that... also seems to work. Right way to do that?

Yes, totally correct, that is what propagates to the consumers, via their generators (CMakeDeps, MSBuildDeps, PkgConfigDeps, AutotoolsDeps, etc), so the library can be correctly consumed by different build systems without users needing to think about the library name. 

> This is also kind of a separate question, but what is then the right way to pass all this same conan profile information (libstdc++ version, c++ standard version, build type) into cmake? I see the generated conanbuildinfo.cmake that I get from running conan install does conditionally try to do the right thing:

It seems that you are referring to the ``generators = "cmake"`` legacy integration. It shouldn't be used, using the ones in https://docs.conan.io/en/latest/reference/conanfile/tools/cmake.html are the recommended way:

- ``CMakeToolchain`` will map the current project settings, like ``compiler.cppstd`` into ``conan_toolchain.cmake`` file. This toolchain file should contain the right flags, also regarding GLIBCXX_USE_CX11
- ``CMakeDeps`` will map the dependencies to ``xxx-config.cmake`` files, so they can be consumed with ``find_package()``

Please have a look to the toolchain file and let us know if it works.

**Comment by memsharded** [comment]
Hi @brevzin 

I am closing this ticket as staled (and also outdated, as it was still referring to some legacy generators and Conan 1)
Don't hesitate to open new tickets for any further question or issue you might have, thanks again for your feedback!
